i3d/src/main/scala/Collateral.scala

```diff
@@ -0,0 +1,440 @@
+        ("path" -> i.pathName.replace(".","/")) ~
+        ("type" -> (if (p.ip.blackBoxModels) "hard_placement" else "hardmacro")) ~
+        ("x" -> roundToNm(i.forBump.location.x - (
+          if (p.ip.blackBoxModels) p.gp.pitchH / 2 else 0.0))) ~
+        ("y" -> roundToNm(i.forBump.location.y - (
+          if (p.ip.blackBoxModels) p.gp.pitchV / 2 else 0.0))) ~
+        ("width" -> (if (p.ip.blackBoxModels) Some(p.gp.pitchH)
+                     else None)) ~
+        ("height" -> (if (p.ip.blackBoxModels) Some(p.gp.pitchV)
+                      else None)) ~
+        ("master" -> (if (p.ip.blackBoxModels) None
+                      else Some(i.desiredName)))
+        // TODO: top layer for halos
+      ) ++ p.flatBumpMap.map( b =>  // Routing KOZ
+        ("path" -> s"Patch/${b.bumpName}_route_koz") ~
+        ("type" -> "obstruction") ~
+        ("obs_types" -> Seq("route", "power")) ~
+        ("x" -> roundToNm(b.location.x - p.ip.viaKOZRatio * p.gp.pitchH / 2)) ~
+        ("y" -> roundToNm(b.location.y - p.ip.viaKOZRatio * p.gp.pitchV / 2)) ~
+        ("width" -> roundToNm(p.ip.viaKOZRatio * p.gp.pitchH)) ~
+        ("height" -> roundToNm(p.ip.viaKOZRatio * p.gp.pitchV))
+      ) ++ (if (placeKOZ > 0) p.flatBumpMap.map( b =>  // Place KOZ
+        ("path" -> s"Patch/${b.bumpName}_place_koz") ~
+        ("type" -> "obstruction") ~
+        ("obs_types" -> Seq("place")) ~
+        ("x" -> roundToNm(b.location.x - placeKOZ * p.gp.pitchH / 2)) ~
+        ("y" -> roundToNm(b.location.y - placeKOZ * p.gp.pitchV / 2)) ~
+        ("width" -> roundToNm(placeKOZ * p.gp.pitchH)) ~
+        ("height" -> roundToNm(placeKOZ * p.gp.pitchV))
+      ) else Seq.empty))) ~
+      ("vlsi.inputs.placement_constraints_meta" -> "append")
+
+    // SDC: clocks, delays, loads
+    val sdc =
+      ("vlsi.inputs.custom_sdc_files" -> Seq("Patch.sdc")) ~
+      ("vlsi.inputs.custom_sdc_files_meta" -> "prependlocal")
+
+    // Power intent
+    val power =
+      ("vlsi.inputs.power_spec_mode" -> "auto") ~
+      ("vlsi.inputs.power_spec_type" -> "upf") ~
+      ("vlsi.inputs.supplies.power" -> Seq(("name" -> "VDDI") ~ ("pins" -> Seq("VDDI")))) ~
+      ("vlsi.inputs.supplies.ground" -> Seq(("name" -> "VSS") ~ ("pins" -> Seq("VSS"))))
+
+    pretty(render(bumps merge pins merge places merge sdc merge power))
+  }
+
+  def toSDC(shiftCase: Boolean = false): String = {
+    // Extract bump objects
+    val bumps = iocells.map(_.forBump)
+
+    // TODO: power is bump pitch related.
+    val globalSDC = f"""# Global constraints
+      |# All timing units assumed in ns
+      |# Loads/transitions (config pins, internal)
+      |set_load 0.005 [all_outputs]
+      |set_max_transition 0.05 [current_design]
+      |# Power. Not supported by all tools.
+      |set_max_dynamic_power ${0.05 * p.numMods * p.sigsPerMod / p.tPeriod}%.1f mW [current_design]
+      |# False paths from config pins
+      |set_false_path -from *ioCtrl*
```

Comment:
I don't like `ioCtrl` as a false path targeting string. That could easily get used for unrelated code.

This string should be very verbose, see the example I showed you the other day!